### PR TITLE
swev-id: django__django-14559 ORM: Return number of rows matched from bulk_update()

### LINKED ISSUE #432: Include number of rows matched in bulk_update() return value
Issue: Include number of rows matched in bulk_update() return value

Description  
Currently, bulk_update() returns None, unlike update(), which returns the number of rows matched.
It looks like it would be easy to add the same functionality to bulk_update() since bulk_update() simply calls update() repeatedly:
https://github.com/django/django/blob/2b4b6c8af0aae8785bc1347cf1be2e8e70fd5ff3/django/db/models/query.py#L568
I.e. the return values could simply be added and returned.

Issue Comments:
Also, if this feature is okayed, maybe the number could be returned in a way that would permit other useful numbers to be added to the return value in a backwards-compatible way later on. For example, the number could be returned as a property of a named tuple with one property. That way, if a new value is requested in the future, it could be added by adding a new property to the named tuple.
Also, if this feature is okayed, maybe the number could be returned in a way that would permit other useful numbers to be added to the return value in a backwards-compatible way later on. For example, the number could be returned as a property of a named tuple with one property. That way, if a new value is requested in the future, it could be added by adding a new property to the named tuple. Do you have any specific statistics in mind?
Thanks. Do you have any specific statistics in mind? Not especially, but more because I don't know what might be available. The only things that occur to me right now are things like the number of queries made, and the batch size used for the function call. More usefully, though, if update() were ever to grow the ability to report not just the number of rows matched but also the number of rows changed, that would definitely be of interest.
More usefully, though, if update() were ever to grow the ability to report not just the number of rows matched but also the number of rows changed, that would definitely be of interest. Given that would likely require a signature change to update if we ever want to support it I think that making bulk_update return the sums of its updates should be good enough for now.
I would second that: it’s good to be future proof, but we’re somewhat locked in to the current returned result due to update. Let’s just make it return the updated rows, which seems like a simple and sensible change.
I took a look at this yesterday and due to the way bulk_update treats duplicates in objs this might not be quite so straight forward as simply summing the returned values from repeated update() calls. Duplicates split between batches will be counted twice. See https://code.djangoproject.com/ticket/32388
There is now an (accepted) issue #32406 to make update() return something different. Given that, might it make sense to reconsider making bulk_update() future-proof now?
update will keep returning int unless returning is specified otherwise it would break backward compatibility. Can't this ticket be dedicated to having bulk_update returning int when returning is not specified? I don't see what needs to be made here to make bulk_update future proof? How does making bulk_update -> int prevent us from having bulk_update(returning) -> list match the signature of update(returning) -> list in the future?
assign to bhavikabadjate901
Due to inactivity, I'm claiming this ticket. bhavikabadjate901 please come forward if you have done any work on this! I'm taking the simplest approach first, which is bulk_update -> int. As stated by Simon Charette: I think that making bulk_update return the sums of its updates should be good enough for now. and Tom Forbes: Let’s just make it return the updated rows Then, I'll stop and think a bit more about Chris Jerdonek's future-proofing and Tim McCurrach's duplicate edge case.
PR ready for review at https://github.com/django/django/pull/14125. I did not handle duplicates, and did not future-proof bulk_update. I'm keeping an eye on #32406. If it gets patched, I'll update the returned value on bulk_update. If we pin down an API that is generally agreed upon, I might claim that ticket as well.

Acceptance criteria:
- bulk_update() returns an int equal to the sum of rows matched by per-batch updates; empty input returns 0.
- Document and accept the limitation that duplicates across batches are double-counted; duplicates within the same batch count once and the first occurrence’s value wins.
- Add tests covering single, multiple, batched updates, no matches, and duplicate scenarios.

Reference: Task request to implement on branch django__django-14559, and open PR titled with token 'swev-id: django__django-14559'.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- aggregate the per-batch counts returned by QuerySet.bulk_update() and return the total rows matched
- document the new return value and add regression coverage for note bulk updates, including duplicate objects
- capture the duplicate caveat: first occurrence wins inside a batch, but duplicates rerun in later batches are double-counted

## Testing
- PYTHONPATH=/workspace/django:$HOME/.nix-profile/lib/python3.11/site-packages DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py queries.test_bulk_update --parallel=1
- flake8 django/db/models/query.py tests/queries/test_bulk_update.py

Resolves #432

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Local Validation
- `PYTHONPATH=/workspace/django:$HOME/.nix-profile/lib/python3.11/site-packages DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py queries.test_bulk_update --parallel=1` (passed; 31 tests, 0 failures)
- `flake8 django/db/models/query.py tests/queries/test_bulk_update.py` (passed; no lint errors)

File: django/db/models/query.py [hunk 1]
```diff
@@ -527,6 +527,7 @@ def bulk_create(self, objs, batch_size=None, ignore_conflicts=False):
     def bulk_update(self, objs, fields, batch_size=None):
         """
         Update the given fields in each of the given objects in the database.
+        Return the number of rows matched (int).
```
Comment (noa-lucent) [inline django/db/models/query.py:530]: [major] Thanks for wiring the return value through the implementation. We also need to update the public docs to describe this API change.  still says nothing about a return value and only mentions that duplicates within a batch keep the first instance. Please document that  now returns the number of matched rows and explicitly call out that duplicates split across batches will cause the total to double count. Without that doc update the change is effectively invisible (and the acceptance criteria from #432 aren't met).}

File: tests/queries/test_bulk_update.py [hunk 2]
```diff
@@ -240,3 +240,73 @@ def test_json_field(self):
...
+
+        count = Note.objects.bulk_update(notes, ['note'], batch_size=1)
+
+        self.assertEqual(count, len(notes))
+        self.assertCountEqual(
+            Note.objects.values_list('note', flat=True),
+            [note.note for note in notes],
+        )
+
+    def test_returns_zero_for_no_matches(self):
```
Comment (noa-lucent) [inline tests/queries/test_bulk_update.py:284]: [minor] Nice coverage on the missing-row case. We also need a regression assertion for the empty-input branch now that  returns 0 there.  currently only checks the query count; please update it (or add a sibling test) to assert the method returns 0 so we lock down that acceptance criterion.}

Comment (noa-lucent) [inline tests/queries/test_bulk_update.py:284]: [minor] Nice coverage on the missing-row case. We also need a regression assertion for the empty-input branch now that bulk_update() returns 0 there. Please update `BulkUpdateTests.test_empty_objects()` (or add a similar test) to assert the method returns 0 so we lock down that acceptance criterion.}

File: django/db/models/query.py (hunk 1 above)
Comment (noa-lucent) [inline django/db/models/query.py:530]: [major] Thanks for wiring the return value through the implementation. We also need to update the public docs to describe this API change. `docs/ref/models/querysets.txt` still says nothing about a return value and only mentions that duplicates within a batch keep the first instance. Please document that `bulk_update()` now returns the number of matched rows and explicitly call out that duplicates split across batches will cause the total to double count. Without that doc update the change is effectively invisible (and the acceptance criteria from #432 aren't met).}

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for wiring the count through bulk_update and for the thorough tests. Before we merge, please document the new behaviour in docs/ref/models/querysets.txt (including the duplicate-across-batches caveat) and add a regression assertion for the empty-input branch returning 0 so we lock down the acceptance criteria.}

**Comment by casey-brooks** [comment]
## Local Validation (update)
- `PYTHONPATH=/workspace/django:$HOME/.nix-profile/lib/python3.11/site-packages DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py queries.test_bulk_update --parallel=1` (passed; 32 tests, 0 failures)
- `flake8 django/db/models/query.py tests/queries/test_bulk_update.py` (passed; no lint errors)

**Review by noa-lucent — APPROVED** [review]
Thanks for the quick turnaround—docs now explain the return value and duplicate caveat, and the empty-input regression test covers the 0 return path. Everything looks good to me.}
